src/basilisp/contrib/nrepl_server.lpy

```diff
@@ -0,0 +1,476 @@
+                  (error :request-handler-unexpected-exception (pr-str e)))))
+            (recur (.recv socket recv-buffer-size)))))
+      (catch python/Exception e
+        (error :client-connection-error :client client-info :exception e)
+        (error (traceback/format_exc))))))
+
+(defn server-make [opts]
+  "Create and return a `socketserver/TCPServer` serving nREPL clients
+  according to ``opts` at 127.0.0.1 at a random available port.
+
+  See `ops` for the operation supported by the server.
+
+  The nREPL starts at the `user` namespace and binds `*1`, `*2`, `*3`
+  and `*e` to the ultimate, penultimate, antepenultimate evaluation
+  result and last exception message respectively.
+
+  ``opts`` is a map of options with the following optional keys
+
+  :host The host address to bind to, defaults to 127.0.0.1.
+
+  :port The port number to listen to, defaults to 0 which means to
+  pickup a random available port.
+
+  See :lpy:fn:`on-connect` for additionally supported ``opts`` keys.
+
+  Known limitations:
+
+  1. All client connections share the same environment at the moment,
+  which is the env that the server runs in. This could change in the
+  future to isolate the clients interactions from each other.
+
+  2. The session uuids are ignored and only created to satisfy the
+  initial clone op."
+  (let [{:keys [host port] :or {host "127.0.0.1" port 0}} opts
+        handler (python/type (name (gensym "mytcphandler")) (python/tuple [socketserver/StreamRequestHandler])
+                             #py {"handle" #(on-connect % opts)})]
+    (socketserver/ThreadingTCPServer (python/tuple [host port]) handler)))
+
+(def nrepl-server-signature
+  "The de facto signature nrepl started message that is used by IDEs to
+  identify the host and port number the server is running on."
+  "nREPL server started on port %s on host %s - nrepl://%s:%s")
+
+(defn start-server!
+  "Create an nREPL server with lpy:fn:`server-make` (of which see)
+  according to ``opts`` if given, and serve clients for ever.
+
+  It prints out the `nrepl-server-signature` message at startup for
+  IDEs to pickup the host number to connect to.
+
+  ``opts`` is a map of options with the following optional keys
+
+  :nrepl-port-file The file to write the port number to, it defaults
+  to .nrepl-port.
+
+  :server* An atom to hold the server reference as returned from
+  lpy:fn:`server-make`, useful for testing.
+
+  also see lpy:fn:`server-make` for additionally supported
+  ``opts``keys."
+  ([]
+   (start-server! {}))
+  ([opts]
+   (let [{:keys [nrepl-port-file server*] :or {nrepl-port-file ".nrepl-port"}} opts
+         server (server-make opts)]
+     (when server* (reset! server* server))
+     (try
+       (let [[host port] (py->lisp (.-server-address server))]
+         (println (format nrepl-server-signature port host host port))
+         (spit nrepl-port-file (str port)))
+       (.serve-forever server)
+       (catch python/Exception e
```

Comment:
If users are creating this in a terminal process, we may want to specially handle `KeyboardInterrupt` and shutdown gracefully rather than failing out with an error.

Reply:
Added a KeyboardInterrupt handler.